Lib/dataclasses.py

```diff
@@ -240,6 +240,20 @@ def __repr__(self):
                 f'metadata={self.metadata}'
                 ')')
 
+    # This is used to support the PEP 487 __set_name__ protocol in the
+    #  case where we're using a field that contains a descriptor as a
+    #  defaul value.  For details on __set_name__, see
+    #  https://www.python.org/dev/peps/pep-0487/#implementation-details.
+    # Note that in _process_class, this Field object is overwritten with
+    #  the default value, so the end result is a descriptor that had
+    #  __set_name__ called on it at the right time.
+    def __set_name__(self, owner, name):
+        func = getattr(self.default, '__set_name__', None)
```

Comment:
I got here late, but shouldn't this check for `__set_name__` on `type(self.default)` instead? That's how dunder protocols usually work, and it's also how typeobject.c itself implements `__set_name__` (using `_PyObject_LookupSpecial`).

Reply:
Good catch, thanks. I've created https://bugs.python.org/issue33175.